### PR TITLE
Tests li wand 2016 loss

### PR DESCRIPTION
**Pull request by jbueltemeier**

### CODE REVIEW

File: tests/li_wand_2016/test_loss.py [hunk 1]
```diff
@@ -1,29 +1,188 @@
...
+
+def test_LiWand2016MRFOperator(
+    subtests, multi_layer_encoder_with_layer, target_image, input_image
+):
+    multi_layer_encoder, layer = multi_layer_encoder_with_layer
+    encoder = multi_layer_encoder.extract_encoder(layer)
+    target_enc = encoder(target_image)
+    input_enc = encoder(input_image)
+
+    configs = ((True, 3, 1, 1.0 / 2.0, "sum"), (False, 3, 1, 1.0, "sum"))
```
Comment (pmeier) [inline tests/li_wand_2016/test_loss.py:60]: Is there a point to include `patch_size`, `stride`, and `loss_reduction` if they do not vary with `impl_params`?

File: tests/li_wand_2016/test_loss.py [hunk 2]
```diff
@@ -1,29 +1,188 @@
...
+            if impl_params:
+                target_repr = loss.extract_normalized_patches2d(
+                    target_enc, patch_size, stride
+                )
+                input_repr = loss.extract_normalized_patches2d(
+                    input_enc, patch_size, stride
+                )
+            else:
+                target_repr = extract_patches2d(target_enc, patch_size, stride)
+                input_repr = extract_patches2d(input_enc, patch_size, stride)
```
Comment (pmeier) [inline tests/li_wand_2016/test_loss.py:82]:
```suggestion
           extract_patches = loss.extract_normalized_patches2d if impl_params else extract_patches2d
           target_repr = extract_patches(target_enc, patch_size, stride)
           input_repr = extract_patches(input_enc, patch_size, stride)
```

File: tests/li_wand_2016/test_loss.py [hunk 3]
```diff
@@ -1,29 +1,188 @@
...
+
+
+def test_li_wand_2016_style_loss(subtests):
+    style_loss = loss.li_wand_2016_style_loss()
+    assert isinstance(style_loss, ops.MultiLayerEncodingOperator)
+
+    with subtests.test("encoding_ops"):
+        assert all(isinstance(op, ops.MRFOperator) for op in style_loss.operators())
+
+    configs = ((True, 1e-4, 2, 1, 5e-2, 1, 7.5), (False, 1e0, 1, 3, 5e-2, 2, 7.5))
```
Comment (pmeier) [inline tests/li_wand_2016/test_loss.py:97]: Same as above for `scale_step_width` and `rotate_step_width`

File: tests/li_wand_2016/test_loss.py [hunk 4]
```diff
@@ -1,29 +1,188 @@
...
+        )
+        rotation_angles *= rotate_step_width
+
+        for i, values in enumerate(itertools.product(scaling_factors, rotation_angles)):
+            scaling_factor, rotation_angle = values
+            with subtests.test("scaling_factor"):
+                assert op_target_transforms[0][i].scaling_factor == scaling_factor
+
+            with subtests.test("rotation_angles"):
+                assert op_target_transforms[0][i].rotation_angle == rotation_angle
```
Comment (pmeier) [inline tests/li_wand_2016/test_loss.py:143]: Not sure about that. I think we should mock `MRFOperator.scale_and_rotate_transforms` and check if its called with the correct parameters.

File: tests/li_wand_2016/test_loss.py [hunk 5]
```diff
@@ -1,29 +1,188 @@
...
+
+
+def test_LiWand2016TotalVariationOperator(subtests, input_image):
+    configs = ((True, "sum", 1.0 / 2.0), (False, "sum", 1.0))
```
Comment (pmeier) [inline tests/li_wand_2016/test_loss.py:147]: Same as above for `loss_reduction`.

File: tests/li_wand_2016/test_loss.py (hunk 4 above)
Comment (jbueltemeier) [reply tests/li_wand_2016/test_loss.py:136]: I think you are right, here you should not check if the transformations are correct, but if this function is called correctly. The actual test should take place in the tests of the `MRFOperator`.

File: tests/li_wand_2016/test_loss.py [hunk 6]
```diff
@@ -1,29 +1,173 @@
...
+
+            op = loss.LiWand2016MRFOperator(encoder, 3, impl_params=impl_params,)
+            op.set_target_image(target_image)
+            actual = op(input_image)
+
+            extract_patches = (
+                loss.extract_normalized_patches2d if impl_params else extract_patches2d
+            )
+            target_repr = extract_patches(target_enc, 3, 1)
+            input_repr = extract_patches(input_enc, 3, 1)
```
Comment (pmeier) [inline tests/li_wand_2016/test_loss.py:69]: 1. Is there a reason you create the `repr`s in every iteration? If not, I suggest we replace the `enc`s above.
2. I think it would make it more clear to create variables for the `patch_size` and `stride` and use them here and in the constructor of the operator.

File: tests/li_wand_2016/test_loss.py [hunk 7]
```diff
@@ -1,29 +1,173 @@
...
+        pyramid_num_scale_steps = kwargs["num_scale_steps"]
+        pyramid_scale_step_width = kwargs["scale_step_width"]
+        pyramid_num_rotate_steps = kwargs["num_rotate_steps"]
+        pyramid_rotate_step_width = kwargs["rotate_step_width"]
+
+        with subtests.test("num_scale_steps"):
+            assert pyramid_num_scale_steps == num_scale_steps
+
+        with subtests.test("num_scale_steps"):
+            assert pyramid_scale_step_width == 5e-2
```
Comment (pmeier) [inline tests/li_wand_2016/test_loss.py:124]:
```suggestion
            assert pyramid_scale_step_width == pytest.approx(5e-2)
```

File: tests/li_wand_2016/test_loss.py [hunk 8]
```diff
@@ -1,29 +1,173 @@
...
+
+        with subtests.test("num_rotate_steps"):
+            assert pyramid_num_rotate_steps == num_rotate_steps
+
+        with subtests.test("rotate_step_width"):
+            assert pyramid_rotate_step_width == 7.5
```
Comment (pmeier) [inline tests/li_wand_2016/test_loss.py:130]:
```suggestion
            assert pyramid_rotate_step_width == pytest.approx(7.5)
```

File: tests/li_wand_2016/test_loss.py [hunk 9]
```diff
@@ -1,29 +1,173 @@
...
+            assert style_loss.score_weight == pytest.approx(score_weight)
+
+        with subtests.test("layer_weights"):
+            assert layer_weights == (1.0,) * len(layers)
+
+        _, kwargs = mock.call_args
+        pyramid_num_scale_steps = kwargs["num_scale_steps"]
+        pyramid_scale_step_width = kwargs["scale_step_width"]
+        pyramid_num_rotate_steps = kwargs["num_rotate_steps"]
+        pyramid_rotate_step_width = kwargs["rotate_step_width"]
```
Comment (pmeier) [inline tests/li_wand_2016/test_loss.py:118]: Maybe we can move this to a separate function like `test_li_wand_2016_style_loss_pyramid` and only check the type here.

File: tests/li_wand_2016/test_loss.py [hunk 10]
```diff
@@ -1,29 +1,173 @@
...
+
+
+def test_LiWand2016TotalVariationOperator(subtests, input_image):
+    configs = ((True, 1.0 / 2.0), (False, 1.0))
+    for impl_params, score_correction_factor in configs:
+        with subtests.test(impl_params=impl_params):
+            op = loss.LiWand2016TotalVariationOperator(impl_params=impl_params,)
+            actual = op(input_image)
+
+            score = total_variation_loss(input_image, exponent=2.0, reduction="sum")
```
Comment (pmeier) [inline tests/li_wand_2016/test_loss.py:140]: With this we are independent if the default value for the `exponent` changes.
```suggestion
            score = total_variation_loss(input_image, exponent=op.exponent, reduction="sum")
```

File: tests/li_wand_2016/test_loss.py (hunk 6 above)
Comment (jbueltemeier) [reply tests/li_wand_2016/test_loss.py:69]: To point 1: `extract_patches` depends on the `impl_params`.

```python
extract_patches = (loss.extract_normalized_patches2d if impl_params else extract_patches2d)
```

File: tests/li_wand_2016/test_loss.py [hunk 11]
```diff
@@ -1,29 +1,180 @@
...
+    input_enc = encoder(input_image)
+
+    patch_size = 3
+    stride = 1
+    configs = ((True, 1.0 / 2.0), (False, 1.0))
+    for (impl_params, score_correction_factor,) in configs:
+        with subtests.test(impl_params=impl_params):
+
+            op = loss.LiWand2016MRFOperator(
+                encoder, patch_size, impl_params=impl_params,
```
Comment (pmeier) [inline tests/li_wand_2016/test_loss.py:64]:
```suggestion
                encoder, patch_size, impl_params=impl_params, stride=stride
```
